cupy/linalg/product.py

```diff
@@ -59,6 +60,135 @@ def vdot(a, b):
+        axisb : int, optional
+            Axis of ``b`` that defines the vector(s).
+            By default, the last axis.
+        axisc : int, optional
+            Axis of ``c`` containing the cross product vector(s).  Ignored if
+            both input vectors have dimension 2, as the return is scalar.
+            By default, the last axis.
+        axis : int, optional
+            If defined, the axis of ``a``, ``b`` and ``c``
+            that defines the vector(s) and cross product(s).
+            Overrides ``axisa``, ``axisb`` and ``axisc``.
+
+    Returns:
+        c : cupy.ndarray
+            Vector cross product(s).
+
+    .. seealso:: :func:``numpy.cross``
+
+    """
+
+    if axis is not None:
+        axisa, axisb, axisc = (axis,) * 3
+    a = cupy.asarray(a)
+    b = cupy.asarray(b)
+    # Check axisa and axisb are within bounds
+    # TODO: Implement normalize_axis_index in cupy
+
+    axisa = chebyshev.normalize_axis_index(
+        axisa, a.ndim, msg_prefix='axisa')
+    axisb = chebyshev.normalize_axis_index(
+        axisb, b.ndim, msg_prefix='axisb')
+
+    # Move working axis to the end of the shape
+    a = cupy.moveaxis(a, axisa, -1)
+    b = cupy.moveaxis(b, axisb, -1)
+    msg = ('incompatible dimensions for cross product\n'
+           '(dimension must be 2 or 3)')
+    if a.shape[-1] not in (2, 3) or b.shape[-1] not in (2, 3):
+        raise ValueError(msg)
+
+    # Create the output array
+    shape = cupy.broadcast(a[..., 0], b[..., 0]).shape
+    if a.shape[-1] == 3 or b.shape[-1] == 3:
+        shape += (3,)
+        # Check axisc is within bounds
+        axisc = chebyshev.normalize_axis_index(
+            axisc, len(shape), msg_prefix='axisc')
+    dtype = cupy.promote_types(a.dtype, b.dtype)
+    cp = cupy.empty(shape, dtype)
+
+    # create local aliases for readability
+    a0 = a[..., 0]
+    a1 = a[..., 1]
+    if a.shape[-1] == 3:
+        a2 = a[..., 2]
+    b0 = b[..., 0]
+    b1 = b[..., 1]
+    if b.shape[-1] == 3:
+        b2 = b[..., 2]
+    if cp.ndim != 0 and cp.shape[-1] == 3:
+        cp0 = cp[..., 0]
+        cp1 = cp[..., 1]
+        cp2 = cp[..., 2]
+
+    if a.shape[-1] == 2:
+        if b.shape[-1] == 2:
+            # a0 * b1 - a1 * b0
+            cupy.multiply(a0, b1, out=cp)
+            cp -= a1 * b0
+            return cp
+        else:
+            assert b.shape[-1] == 3
+            # cp0 = a1 * b2 - 0  (a2 = 0)
+            # cp1 = 0 - a0 * b2  (a2 = 0)
+            # cp2 = a0 * b1 - a1 * b0
+            cupy.multiply(a1, b2, out=cp0)
+            cupy.multiply(a0, b2, out=cp1)
+            cupy.negative(cp1, out=cp1)
+            cupy.multiply(a0, b1, out=cp2)
+            cp2 -= a1 * b0
+    else:
+        assert a.shape[-1] == 3
+        if b.shape[-1] == 3:
+            # cp0 = a1 * b2 - a2 * b1
+            # cp1 = a2 * b0 - a0 * b2
+            # cp2 = a0 * b1 - a1 * b0
+            cupy.multiply(a1, b2, out=cp0)
+            tmp = cupy.array(a2 * b1)
```

Comment:
I think you don't have to apply `cupy.array()`.

Reply:
`cp0`, `cp1`, `cp2` are all arrays and they get subtracted from `tmp`, so its always safer to use `tmp` as an `cupy.array()`

Reply:
I mean `a2` and `b1` are always `cupy.ndarray`s. So `a2 * b1` is always `cupy.ndarray`, too.

Reply:
Yes I have added the changes